### DISCUSSION TITLE
Git branch integration

### DISCUSSION THREAD

**Discussion by tux-84** [ankitpokhrel/jira-cli#711] [Ideas]
An idea, which the (currently broken again) VSC plugin features, is creating a gut branch based on an issue key when starting to work on it. Maybe this could be realized the following way:

1.  Command `jira issue start (<issue-key>)`
2. Dialog pops up for issue selection unless given by previous command
3. Plugin runs:
- git branch `KEY-1234' 
- git push origin HEAD:KEY-1234

**Comment by addynaik** [comment]
This is easily setup through git commands - 

here are the steps: (on a mac and linx - someone can provide similar for windows)
* create a folder in home called `~/git-scripts`
* create a file called `git-start`
* make the file executable `chmod +x git-start`
* add the following content
```
#!/bin/sh

type=$1
ticket=$2
test -z "$ticket" && echo "ERROR: Please provide the jira ticket number" 1>&2 && exit 1

jiraBranchName=$(jira issue list -q "key = $ticket" --plain | awk 'NR==2' | awk -F '\t' '{ print $3 }')
echo ""
test -z "$jiraBranchName" && echo "ERROR: Please provide valid jira ticket number" 1>&2 && exit 1

branchName=$(echo "$jiraBranchName" | sed "s/[^[:alpha:].-]/-/g" | tr '[:upper:]' '[:lower:]' | head -c 30)

echo starting $type for "\"$jiraBranchName\"" - "$type/$ticket-$branchName"
# Stash current changes
git stash
# Checkout remote branch from where you want to update. 
git fetch
git checkout develop
git pull origin develop
git co -b "$type/$ticket-$branchName"
```
* add the following to your `.bashrc` or `.zshrc` - `export PATH="$PATH:$HOME/git-scripts"`
* now you can run the command `git start feature KEY-1234` or `git start fix KEY-1234`

sample output:
```
 ~/code/data-intergration/data-factory - develop - git start feature KEY-326

starting feature for "Add Snowflake Table" - feature/KEY-326-add-snowflake-table
No local changes to save
Already on 'develop'
Your branch is up to date with 'origin/develop'.
From XXXXXXXXXXXXX
 * branch            develop    -> FETCH_HEAD
Already up to date.
Switched to a new branch 'feature/KEY-326-add-snowflake-table'
 ~/code/data-intergration/data-factory - feature/KEY-326-add-snowflake-table - 
```

**Reply by benceferdinandy-signifyd** [reply]
It would be more agnostic if it used `origin/HEAD` instead of develop and with `git switch -c "$branchName" origin/HEAD --no-track` it will cleanly create a new branch from whatever is upstream without needing to first check and the pull anything to the local develop/main/master branch.

**Comment by neugartf** [comment]
Can plug this cli tool here which somewhat achieves something similar :). 
https://github.com/freenowtech/git-jira
